src/ray/util/logging.cc

```diff
@@ -0,0 +1,147 @@
+#include <cstdlib>
+#include <iostream>
+
+#include "ray/util/logging.h"
+
+#ifdef RAY_USE_GLOG
+#include "glog/logging.h"
+#endif
+
+namespace ray {
+
+// This is the default implementation of ray log,
+// which is independent of any libs.
+class CerrLog {
+ public:
+  CerrLog(int severity) : severity_(severity), has_logged_(false) {}
+
+  virtual ~CerrLog() {
+    if (has_logged_) {
+      std::cerr << std::endl;
+    }
+    if (severity_ == RAY_FATAL) {
+      PrintBackTrace();
+      std::abort();
+    }
+  }
+
+  std::ostream &Stream() {
+    has_logged_ = true;
+    return std::cerr;
+  }
+
+  template <class T>
+  CerrLog &operator<<(const T &t) {
+    if (severity_ != RAY_DEBUG) {
+      has_logged_ = true;
+      std::cerr << t;
+    }
+    return *this;
+  }
+
+ protected:
+  const int severity_;
+  bool has_logged_;
+
+  void PrintBackTrace() {
+#if defined(_EXECINFO_H) || !defined(_WIN32)
+    void *buffer[255];
+    const int calls = backtrace(buffer, sizeof(buffer) / sizeof(void *));
+    backtrace_symbols_fd(buffer, calls, 1);
+#endif
+  }
+};
+
+int RayLog::severity_threshold_ = RAY_INFO;
+
+#ifdef RAY_USE_GLOG
+using namespace google;
+
+// Glog's severity map.
+static int GetMappedSeverity(int severity) {
+  static int severity_map[] = {
+      // glog has no DEBUG level. It has verbose level but hard to adapt here.
+      GLOG_INFO,     // RAY_DEBUG
+      GLOG_INFO,     // RAY_INFO
+      GLOG_WARNING,  // RAY_WARNING
+      GLOG_ERROR,    // RAY_ERROR
+      GLOG_FATAL     // RAY_FATAL
+  };
+  // Ray log level starts from -1 (RAY_DEBUG);
+  return severity_map[severity + 1];
+};
+
+#endif
+
+void RayLog::StartRayLog(const std::string &app_name, int severity_threshold,
+                         const std::string &log_dir) {
+#ifdef RAY_USE_GLOG
+  std::string dir_ends_with_slash = std::string(log_dir);
+  if (!dir_ends_with_slash.empty()) {
+    if (dir_ends_with_slash[dir_ends_with_slash.length() - 1] != '/') {
+      dir_ends_with_slash += "/";
+    }
+  }
+  std::string app_name_str = app_name;
+  if (app_name_str.empty()) {
+    app_name_str = "DefaultApp";
+  } else {
+    // Find the app name without the path.
```

Comment:
I didn't look closely.  What if we just do `google:: InitGoogleLogging(argv[0])`?  What's the behavior difference between just doing that one line and this function?

Reply:
+1 on this suggestion

Reply:
It is fine to pass an empty name to `google:: InitGoogleLogging`. So I changed the logic. `app_name_without_path` will be used in log file name.

Reply:
Current log lib is pluggable, so we need to use `RayLog::StartRayLog` instead of `google:: InitGoogleLogging`. This enable ray to extend the log with other libs.